### PR TITLE
Add Haskell IDE Engine (hie) support

### PR DESCRIPTION
**Pull request by Luxed**
This PR should address the issue #1578.

I don't think it is ready to merge yet.
For me, hie is working in a stack project or on a single file. I don't know about cabal projects (it shouldn't because it only searches for "stack.yaml", maybe I should search for the nearest "*.cabal" file too ?).

Help is appreciated :beer:.

TODO list:
- [x] Make a function to find the nearest file corresponding to `*.cabal`
- [x] Do more testing on plain GHC project (no file to identify them)
- [x] Add tests for hie

### CODE REVIEW

File: ale_linters/haskell/hie.vim [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    let l:executable = ale_linters#haskell#hie#GetExecutable(a:buffer)
+    let l:arguments = '--lsp'
+
+    return ale#Escape(l:executable) . ' ' . l:arguments
+endfunction
+
+function! ale_linters#haskell#hie#GetProjectRoot(buffer) abort
+    let l:stack_file = ale#path#FindNearestFile(a:buffer, 'stack.yaml')
+
+    return !empty(l:stack_file) ? fnamemodify(l:stack_file, ':h') : getcwd()
```
Comment (w0rp) [inline ale_linters/haskell/hie.vim]: If you can't find a project file, return `''` and the language server won't be started. You can only use a language server connection for projects. As you say, it's worth looking for Cabal files too.

File: ale_linters/haskell/hie.vim [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+function! ale_linters#haskell#hie#GetExecutable(buffer) abort
+    return ale#Var(a:buffer, 'haskell_hie_executable')
+endfunction
+
+function! ale_linters#haskell#hie#GetCommand(buffer) abort
+    let l:executable = ale_linters#haskell#hie#GetExecutable(a:buffer)
+    let l:arguments = '--lsp'
+
+    return ale#Escape(l:executable) . ' ' . l:arguments
+endfunction
```
Comment (w0rp) [inline ale_linters/haskell/hie.vim]: Try writing just `'command': '%e --lsp',` below. I added support for a special `%e` format marker for adding in the escaped executable recently.

File: ale_linters/haskell/hie.vim (hunk 1 above)
Comment (w0rp) [reply ale_linters/haskell/hie.vim]: Make sure not to use `getcwd()`. See `:help ale-development` for lots of information, including why `getcwd()` is banned.

**Comment by Luxed** [comment]
Everything should be good now !
This is how a project is found: it searches for a `stack.yaml`, if it doesn't find it, it will search for a `*.cabal` file and if it fails again, it will use the buffers directory.

**Comment by w0rp** [comment]
Cheers! :beers:

**Comment by Luxed** [comment]
Cheers! :beers:

**Pull request merged**